library/src/main/java/com/datatorrent/lib/state/managed/BucketsFileSystem.java

```diff
@@ -0,0 +1,573 @@
+
+    public long getSizeInBytes()
+    {
+      return this.sizeInBytes;
+    }
+
+    public long getBucketId()
+    {
+      return bucketId;
+    }
+
+    public long getTimeBucketId()
+    {
+      return timeBucketId;
+    }
+
+    public Slice getFirstKey()
+    {
+      return firstKey;
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+      if (this == o) {
+        return true;
+      }
+      if (!(o instanceof TimeBucketMeta)) {
+        return false;
+      }
+
+      TimeBucketMeta that = (TimeBucketMeta)o;
+
+      return bucketId == that.bucketId && timeBucketId == that.timeBucketId;
+    }
+
+    @Override
+    public int hashCode()
+    {
+      return Objects.hash(bucketId, timeBucketId);
+    }
+
+    @Override
+    public int compareTo(@NotNull TimeBucketMeta o)
+    {
+      if (bucketId < o.bucketId) {
+        return -1;
+      }
+      if (bucketId > o.bucketId) {
+        return 1;
+      }
+      if (timeBucketId < o.timeBucketId) {
+        return -1;
+      }
+      if (timeBucketId > o.timeBucketId) {
+        return 1;
+      }
+      return 0;
+    }
+  }
+
+  /**
+   * Represents time bucket meta information which can be changed.
+   * The updates to an instance and read/creation of {@link #immutableTimeBucketMeta} belonging to it are synchronized
+   * as different threads are updating and reading from it.<br/>
+   *
+   * The instance is updated when data from window files are transferred to bucket files and
+   * {@link com.datatorrent.lib.state.managed.Bucket.DefaultBucket} reads the immutable time bucket meta.
+   */
+  static class MutableTimeBucketMeta extends TimeBucketMeta
+  {
+    private transient ImmutableTimeBucketMeta immutableTimeBucketMeta;
+
+    private volatile boolean changed;
+
+    MutableTimeBucketMeta()
+    {
+      //for kryo
+    }
+
+    public MutableTimeBucketMeta(long bucketId, long timeBucketId)
+    {
+      super(bucketId, timeBucketId);
+    }
+
+    synchronized void updateTimeBucketMeta(long lastTransferredWindow, long bytes, @NotNull Slice firstKey)
+    {
+      changed = true;
+      this.lastTransferredWindowId = lastTransferredWindow;
+      this.sizeInBytes = bytes;
+      this.firstKey = Preconditions.checkNotNull(firstKey, "first key");
+    }
+
+    synchronized ImmutableTimeBucketMeta getImmutableTimeBucketMeta()
+    {
+      if (immutableTimeBucketMeta == null || changed) {
+        immutableTimeBucketMeta = new ImmutableTimeBucketMeta(getBucketId(), getTimeBucketId(),
+            getLastTransferredWindowId(), getSizeInBytes(), getFirstKey());
+        changed = false;
+      }
+      return immutableTimeBucketMeta;
+    }
+
+  }
+
+  /**
+   * This serves the readers - {@link com.datatorrent.lib.state.managed.Bucket.DefaultBucket}.
+   * It is accessible outside the package unlike {@link MutableTimeBucketMeta} that can be only modified by
+   * BucketPersistor.
+   */
+  public static final class ImmutableTimeBucketMeta extends TimeBucketMeta
```

Comment:
Do we need this class ?
`TimeBucketMeta` itself seems immutable since it has no setters and private constructors, so could we not just use that instead ?